### DISCUSSION TITLE
NUCLEO-F746ZG to STM32745VG struggling migration

### DISCUSSION THREAD

**Discussion by YegorB** [micropython/micropython#15273] [STM32 / Pyboard]
Hello community. 
I am an amateur in MCU and I have managed some MP code on Nucleo and started to migrate it to the custom board based on STM32745VG.
First issue|challenge is that target board has only SWD so I could not read any output from the serial port.
Second issue|challenge is that target board has absolutely different pins mapping comparing to NUCLEO.

Steps I've done so far:
1. created new folder STM32F745VG under boards as well as files stm32f745_af.csv and stm32f745.ld based on NUCLEO-746
2. updated pins in pins.csv and left only
```csv
LED1,PD15
LED2,PC7
LED3,PA11
```
3. updated mpconfigboard.h and left only
```c
#define MICROPY_HW_BOARD_NAME       "STM32F745VG"
#define MICROPY_HW_MCU_NAME         "STM32F745"

#define MICROPY_HW_CLK_PLLM (4)
#define MICROPY_HW_CLK_PLLN (216)
#define MICROPY_HW_CLK_PLLP (RCC_PLLP_DIV2)
#define MICROPY_HW_CLK_PLLQ (9)

#define MICROPY_HW_FLASH_LATENCY    FLASH_LATENCY_7 // 210-216 MHz needs 7 wait states

// LEDs
#define MICROPY_HW_LED1             (pin_D15) // green
#define MICROPY_HW_LED2             (pin_C7) // yellow
#define MICROPY_HW_LED3             (pin_A11) // red
#define MICROPY_HW_LED_ON(pin)      (mp_hal_pin_high(pin))
#define MICROPY_HW_LED_OFF(pin)     (mp_hal_pin_low(pin))
```
4. have no ability to update main.py through REPL (or at least I don't see how to do this using just SWD) so created file post_main.py in /modules to include it into firmware as freeze module and put some simple code there:
```py
import pyb

led = pyb.Pin('D15', pyb.Pin.OUT_PP)

while True:
    led.high()
    pyb.delay(1000)
    led.low()
    pyb.delay(1000)
```
5. included such line of code in main.c:625 (based on suggestion [here](https://github.com/orgs/micropython/discussions/14484#discussioncomment-9441618) )
```c
pyexec_frozen_module("post_main.py", true);
```
But nothing happens after I flashed it out. 
Please help me in advance or point me to the documentation if I missed. 

PS. just to be sure board is alive I tried such sample through cubeIDE and it blinks:
```c
while (1)
{
    HAL_GPIO_WritePin(GPIOD, GPIO_PIN_15, GPIO_PIN_SET);
    HAL_Delay(100);
    HAL_GPIO_WritePin(GPIOD, GPIO_PIN_15, GPIO_PIN_RESET);
    HAL_Delay(100);
}
```

**Comment by davefes** [comment]
In your `mpconfigboard.h` have you got anything like:
```py
// UART buses
#define MICROPY_HW_UART1_TX                     (pin_B6)    // Arduino D1, pin 7 on CN9
#define MICROPY_HW_UART1_RX                     (pin_B7)    // Arduino D0, pin 8 on CN9
#define MICROPY_HW_LPUART1_TX                   (pin_A2)    // ST-link
#define MICROPY_HW_LPUART1_RX                   (pin_A3)    // ST-link
#define MICROPY_HW_UART_REPL                    PYB_LPUART_1
#define MICROPY_HW_UART_REPL_BAUD               115200
```

I am trying to get a STM32WL5x board to run.  There seems to be an issue on some of these variants with even trying to upload your .py files using the UART.  

I have had a quick look-around as I too wonder if SWD could be used to upload your .py files.

**Reply by YegorB** [reply]
Hi @davefes. I removed all the configuration for UART in `mpconfigboard.h` as I didn't have access to it. And the only way I see any `.py` files was through the freeze modules... but looks like it will be much easy to do it using UART...

**Reply by davefes** [reply]
OK, it is so much nicer being able to work on your `.py` files using `rsusb` or `mpremote`.  How do developers work on the STM32745VG?

**Reply by YegorB** [reply]
I suppose it is not a question of MCU, but the board specs. I used to use `NUCLEO-F746ZG` which is `STM32746ZG` and it has USB on a board. So had no issues at all using `Thonny` to update all my `.py` files. But on my another board there is no serial ports at all - so ti is really hard to figure out what happened in runtime. 
Also I am interested in MP configuration for the board from scratch - to better understand all the options.

**Comment by bikeNomad** [comment]
You deleted far too much from `mpconfigboard.h`. 

Have you edited the `.ld` file to reflect the different RAM/Flash configuration? And changed the alternate-function table as appropriate for your chip?

You will want storage for your programs. So add:
```
#define MICROPY_HW_HAS_FLASH        (1)
```

The SWD is fine for loading your binary into flash, but you will want some kind of REPL for development. The stock port uses a UART for the REPL. So you will want to define a UART for the REPL (here's the setup for using UART3; change pins and UART number as necessary):
```
#define MICROPY_HW_UART3_TX         (pin_D8)
#define MICROPY_HW_UART3_RX         (pin_D9)
#define MICROPY_HW_UART_REPL        PYB_UART_3
#define MICROPY_HW_UART_REPL_BAUD   11520
```

If you provide connection to a USB port you can have the board appear as a USB MSC storage device. You would need to add definitions for the USB like this (change pins as needed):
```
#define MICROPY_HW_ENABLE_USB       (1)
// USB config (CN13 - USB OTG FS)
#define MICROPY_HW_USB_FS              (1)
#define MICROPY_HW_USB_VBUS_DETECT_PIN (pin_A9)
#define MICROPY_HW_USB_OTG_ID_PIN      (pin_A10
```

**Reply by YegorB** [reply]
Hi @bikeNomad. Thank you for the updates. 

> Have you edited the .ld file to reflect the different RAM/Flash configuration?

No I did nothing in `.ld` file for memory as it has same series as 746, but I'm going to double check.

> The SWD is fine for loading your binary into flash

that is the only connection I have out of the box for this board. however I've found that I could weld pin to `UART7_TX` and double check (hope I will not break anything on the board )))). in that case is it OK to configure it as you mentioned?
```c
#define MICROPY_HW_UART7_TX         (pin_D8)
#define MICROPY_HW_UART7_RX         (pin_D9)
#define MICROPY_HW_UART_REPL        PYB_UART_7
#define MICROPY_HW_UART_REPL_BAUD   11520
```

**Reply by bikeNomad** [reply]
Assuming that those are the right pins, that should be OK for seeing progress.

Add `main.py` to your `manifest.py` file if you haven't already, and put code in there that prints something.

**Reply by markusand** [reply]
Did you achieve to make it work? We're also stuck on compiling for the STM32F745

**Reply by YegorB** [reply]
@markusand unfortunately not - I was stuck there and put it on hold with Micropython. So I made a quick firmware on C. but I'll be back to work on this next month - so let me know if you need to collaborate.

**Reply by markusand** [reply]
We made it work. Not sure if everything works correctly, but at least we achieved REPL and flash drive.
This is how we've got it set

```c
// UART config
#define MICROPY_HW_UART1_TX         (pin_B6)
#define MICROPY_HW_UART1_RX         (pin_B7)
#define MICROPY_HW_UART2_TX         (pin_D5)
#define MICROPY_HW_UART2_RX         (pin_D6)
#define MICROPY_HW_UART3_TX         (pin_D8)
#define MICROPY_HW_UART3_RX         (pin_D9)
#define MICROPY_HW_UART4_TX         (pin_C10)
#define MICROPY_HW_UART4_RX         (pin_C11)
#define MICROPY_HW_UART6_TX         (pin_C6)
#define MICROPY_HW_UART6_RX         (pin_C7)
#define MICROPY_HW_UART7_TX         (pin_E8)
#define MICROPY_HW_UART7_RX         (pin_E7)
#define MICROPY_HW_UART7_RTS        (pin_E9)
#define MICROPY_HW_UART7_CTS        (pin_E10)
#define MICROPY_HW_UART8_TX         (pin_E1)
#define MICROPY_HW_UART8_RX         (pin_E0)

#define MICROPY_HW_UART_REPL        PYB_UART_7
#define MICROPY_HW_UART_REPL_BAUD   115200

// USB config
#define MICROPY_HW_USB_FS              (1)
#define MICROPY_HW_USB_OTG_ID_PIN      (pin_A10)
```

**Reply by YegorB** [reply]
this is awesome. did you make it work based on STM32745VG or another board setup?

**Reply by markusand** [reply]
Used the STM32745VG as base, and then fiddling with pins with info found here and there. Maybe there's more than needed, and maybe there's things that don't work, but at this point is just try and error and optimize things.
